dns/client.go

```diff
@@ -102,13 +148,141 @@ func (r *Client) Resolve(
+// options (e.g. blockHostnames) are honored. It falls back to the system's
+// default resolver otherwise.
 func (r *Client) Lookup(ctx context.Context, hostname string) ([]string, error) {
-	ips, err := net.DefaultResolver.LookupHost(ctx, hostname)
+	vustate := r.vu.State()
+	if vustate == nil {
+		return nil, common.NewInitContextError("using lookup in the init context is not supported")
+	}
+
+	// If k6's dialer is available, honor its blocked hostnames configuration.
+	if d, ok := vustate.Dialer.(*netext.Dialer); ok && d != nil && d.BlockedHostnames != nil {
+		normalized := strings.TrimSuffix(strings.ToLower(hostname), ".")
+		if _, blocked := d.BlockedHostnames.Contains(normalized); blocked {
+			return nil, &Error{
+				Name:    "BlockedHostname",
+				Message: fmt.Sprintf("blocked hostname: %s", normalized),
+				Kind:    Refused,
+			}
+		}
+	}
+
+	if vustate.Dialer == nil {
+		return nil, fmt.Errorf("unable to perform DNS lookup; reason: k6 internal network dialer is not available")
+	}
+	resolver := &net.Resolver{
+		Dial: vustate.Dialer.DialContext,
+	}
+
+	ips, err := resolver.LookupHost(ctx, hostname)
 	if err != nil {
-		return nil, fmt.Errorf("lookup of %s failed: %w", hostname, err)
+		return nil, &Error{
+			Name:    "LookupFailed",
+			Message: fmt.Sprintf("lookup of %s failed: %s", hostname, err.Error()),
+			Kind:    Refused,
+		}
 	}
 
 	return ips, nil
 }
+
+// k6DNSClient wraps dns.Client to use k6's dialer
+// This ensures k6's networking options (blockHostnames, blacklistIPs) are respected
+type k6DNSClient struct {
+	dns.Client
+	k6Dialer lib.DialContexter
+}
+
+// ExchangeContext overrides the default ExchangeContext to use k6's dialer
+func (c *k6DNSClient) ExchangeContext(
+	ctx context.Context,
+	m *dns.Msg,
+	address string,
+) (*dns.Msg, time.Duration, error) {
+	// If k6 dialer is not available, fall back to standard DNS client behavior
+	if c.k6Dialer == nil {
+		return c.Client.ExchangeContext(ctx, m, address)
+	}
+
+	start := time.Now()
+
+	// Create a connection using k6's dialer
+	conn, err := c.k6Dialer.DialContext(ctx, "udp", address)
+	if err != nil {
+		return nil, 0, err
+	}
+	defer func() {
+		closeErr := conn.Close()
+		if closeErr != nil {
+			log.Fatalf("failed to close k6 DNS connection: %v", closeErr)
+		}
+	}()
+
+	// Set a reasonable deadline for the operation
+	var deadlineErr error
+	if deadline, ok := ctx.Deadline(); ok {
+		deadlineErr = conn.SetDeadline(deadline)
+	} else {
+		deadlineErr = conn.SetDeadline(time.Now().Add(5 * time.Second))
+	}
+	if deadlineErr != nil {
+		return nil, 0, fmt.Errorf("unable to set dns connection deadline; reason: %w", err)
+	}
+
+	// Pack the DNS message and write it
+	data, err := m.Pack()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	_, err = conn.Write(data)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Read the response
+	buffer := make([]byte, 4096)
+	n, err := conn.Read(buffer)
+	if err != nil {
+		return nil, 0, err
+	}
```

Comment:
If I understand right those are errors visualized from the user. Should we wrap them to explain better which step is throwing it?

Reply:
Good catch indeed, expanding the error message 👀 🙇🏻